internal/pkg/graph/graph.go

```diff
@@ -215,3 +222,228 @@ func TopologicalOrder[V comparable](digraph *Graph[V]) (*TopologicalSorter[V], e
+	lg.lock.Lock()
+	defer lg.lock.Unlock()
+	lg.status[vertex] = status
+}
+
+// getStatus gets the status of a vertex.
+func (lg *LabeledGraph[V]) getStatus(vertex V) string {
+	lg.lock.Lock()
+	defer lg.lock.Unlock()
+	return lg.status[vertex]
+}
+
+// getLeaves returns the leaves of a given vertex.
+func (lg *LabeledGraph[V]) leaves() []V {
+	lg.lock.Lock()
+	defer lg.lock.Unlock()
+	var leaves []V
+	for vtx := range lg.vertices {
+		if len(lg.vertices[vtx]) == 0 {
+			leaves = append(leaves, vtx)
+		}
+	}
+	return leaves
+}
+
+// getParents returns the parent vertices (incoming edges) of vertex.
+func (lg *LabeledGraph[V]) parents(vtx V) []V {
+	lg.lock.Lock()
+	defer lg.lock.Unlock()
+	var parents []V
+	for v, neighbors := range lg.vertices {
+		if neighbors[vtx] {
+			parents = append(parents, v)
+		}
+	}
+	return parents
+}
+
+// getChildren returns the child vertices (outgoing edges) of vertex.
+func (lg *LabeledGraph[V]) children(vtx V) []V {
+	lg.lock.Lock()
+	defer lg.lock.Unlock()
+	return lg.Neighbors(vtx)
+}
+
+// filterParents filters parents based on the vertex status.
+func (lg *LabeledGraph[V]) filterParents(vtx V, status string) []V {
+	parents := lg.parents(vtx)
+	var filtered []V
+	for _, parent := range parents {
+		if lg.getStatus(parent) == status {
+			filtered = append(filtered, parent)
+		}
+	}
+	return filtered
+}
+
+// filterChildren filters children based on the vertex status.
+func (lg *LabeledGraph[V]) filterChildren(vtx V, status string) []V {
+	children := lg.children(vtx)
+	var filtered []V
+	for _, child := range children {
+		if lg.getStatus(child) == status {
+			filtered = append(filtered, child)
+		}
+	}
+	return filtered
+}
+
+/*
+UpwardTraversal performs an upward traversal on the graph starting from leaves (nodes with no children)
+and moving towards root nodes (nodes with children).
+It applies the specified process function to each vertex in the graph, skipping vertices with the
+"adjacentVertexSkipStatus" status, and continuing traversal until reaching vertices with the "requiredVertexStatus" status.
+The traversal is concurrent and may process vertices in parallel.
+Returns an error if the traversal encounters any issues, or nil if successful.
+*/
+func (lg *LabeledGraph[V]) UpwardTraversal(ctx context.Context, processVertexFunc func(context.Context, V) error, adjacentVertexSkipStatus, requiredVertexStatus string) error {
+	traversal := &graphTraversal[V]{
+		mu:                       sync.Mutex{},
+		seen:                     make(map[V]struct{}),
+		findBoundaryVertices:     func(lg *LabeledGraph[V]) []V { return lg.leaves() },
+		findAdjacentVertices:     func(lg *LabeledGraph[V], v V) []V { return lg.parents(v) },
+		filterVerticesByStatus:   func(g *LabeledGraph[V], v V, status string) []V { return g.filterChildren(v, status) },
+		requiredVertexStatus:     requiredVertexStatus,
+		adjacentVertexSkipStatus: adjacentVertexSkipStatus,
```

Comment:
```suggestion
		findStartVertices:     func(lg *LabeledGraph[V]) []V { return lg.leaves() },
		findNextVertices:     func(lg *LabeledGraph[V], v V) []V { return lg.parents(v) },
		filterPreviousVerticesByStatus:   func(g *LabeledGraph[V], v V, status string) []V { return g.filterChildren(v, status) },
		requiredVertexStatus:     requiredVertexStatus,
		nextVertexSkipStatus: nextVertexSkipStatus,
```
Is this clearer :thinking: i find `adjacent` and `boundary` a little bit ambiguous.

Reply:
That makes sense 👍